clients/drcachesim/tracer/tracer.cpp

```diff
@@ -211,16 +219,185 @@ static uint     tls_offs;
 static int      tls_idx;
 #define TLS_SLOT(tls_base, enum_val) (void **)((byte *)(tls_base)+tls_offs+(enum_val))
 #define BUF_PTR(tls_base) *(byte **)TLS_SLOT(tls_base, MEMTRACE_TLS_OFFS_BUF_PTR)
-/* We leave a slot at the start so we can easily insert a header entry */
-#define BUF_HDR_SLOTS 1
+/* We leave some space at the start so we can easily insert a header entry */
 static size_t buf_hdr_slots_size;
 
+static inline byte *
+atomic_pipe_write(thread_id_t tid, byte *pipe_start, byte *pipe_end);
+
+static inline void
+init_buf_entry(buf_entry_t *entry)
+{
+    entry->data = NULL;
+    entry->size = 0;
+    entry->tid  = 0;
+}
+
+static void
+free_buf_entry(void *entry)
+{
+    dr_global_free(entry, sizeof(buf_entry_t));
+}
+
+static inline void
+circular_buf_enqueue(void *data, ssize_t size, thread_id_t tid)
+{
+    bool done = false;
+    buf_entry_t *entry;
+    do {
+        /* quick racy check without locking */
+        while (((queue_put + 1) & queue_size_mask) == queue_get) {
```

Comment:
You said you saw major performance issues involving contention on the queue: please add comments about that so we know this code needs improvement.